cmake/OpenCVDownload.cmake

```diff
@@ -37,6 +37,54 @@ file(WRITE "${OPENCV_DOWNLOAD_LOG}" "#use_cache \"${OPENCV_DOWNLOAD_PATH}\"\n")
 file(REMOVE "${OPENCV_DOWNLOAD_WITH_CURL}")
 file(REMOVE "${OPENCV_DOWNLOAD_WITH_WGET}")
 
+set(OPENCV_DOWNLOAD_MIRROR_ID "" CACHE STRING "Available mirror ID: gitcode, custom")
+
+function(ocv_init_download)
+  set(__HOST "unknown") # unknown by default
+  if("m${OPENCV_DOWNLOAD_MIRROR_ID}" STREQUAL "mgitcode")
+    set(__HOST "gitcode")
+  elseif("m${OPENCV_DOWNLOAD_MIRROR_ID}" STREQUAL "mcustom")
+    set(__HOST "custom")
+  else()
+    # Run `git remote get-url origin` to get remote source
+    execute_process(
+      COMMAND
+        git remote get-url origin
+      WORKING_DIRECTORY
+        ${CMAKE_SOURCE_DIR}
+      RESULT_VARIABLE
+        RESULT_STATUS
+      OUTPUT_VARIABLE
+        OCV_GIT_ORIGIN_URL_OUTPUT
+      ERROR_QUIET
+    )
+    # if non-git, OCV_GIT_ORIGIN_URL_OUTPUT is empty
+    if(NOT OCV_GIT_ORIGIN_URL_OUTPUT)
+      message(STATUS "ocv_init_download: OpenCV source tree is not fetched as git repository. 3rdparty resources will be downloaded from github.com by default.")
+      return()
+    else()
+      # Check if git origin is github.com
+      string(FIND "${OCV_GIT_ORIGIN_URL_OUTPUT}" "github.com" _found_github)
+      if(NOT ${_found_github} EQUAL -1)
+        set(__HOST "github")
+      endif()
+      # Check if git origin is gitcode.net
+      string(FIND "${OCV_GIT_ORIGIN_URL_OUTPUT}" "gitcode.net" _found_gitcode)
+      if(NOT ${_found_gitcode} EQUAL -1)
+        set(__HOST "gitcode")
+      endif()
+    endif()
+  endif()
+
+  if(__HOST STREQUAL "unknown")
+    message(WARNING "ocv_init_download: Unknown host detected. Using github.com to download 3rdparty components.")
+  elseif(NOT __HOST STREQUAL "github")
+    message(STATUS "ocv_init_download: Using ${__HOST}-hosted mirror to download 3rdparty components.")
+    ocv_cmake_hook_append(OCV_DOWNLOAD_MIRROR_${__HOST} "${CMAKE_CURRENT_SOURCE_DIR}/cmake/mirrors/${__HOST}.cmake")
```

Comment:
Need to check if we have file

Reply:
In the latest commit, either "gitcode.cmake" or "custom.cmake" can be added into hooks. So I think it is not necessary to check whether those files exist.